http-client/Data/KeyedPool.hs

```diff
@@ -0,0 +1,316 @@
+-- each resource was created.
+data PoolList a
+    = One a {-# UNPACK #-} !UTCTime
+    | Cons
+        a
+
+        -- size of the list from this point and on
+        {-# UNPACK #-} !Int
+
+        {-# UNPACK #-} !UTCTime
+        !(PoolList a)
+    deriving F.Foldable
+
+plistToList :: PoolList a -> [(UTCTime, a)]
+plistToList (One a t) = [(t, a)]
+plistToList (Cons a _ t plist) = (t, a) : plistToList plist
+
+plistFromList :: [(UTCTime, a)] -> Maybe (PoolList a)
+plistFromList [] = Nothing
+plistFromList [(t, a)] = Just (One a t)
+plistFromList xs =
+    Just . snd . go $ xs
+  where
+    go [] = error "plistFromList.go []"
+    go [(t, a)] = (2, One a t)
+    go ((t, a):rest) =
+        let (i, rest') = go rest
+            i' = i + 1
+         in i' `seq` (i', Cons a i t rest')
+
+-- | Create a new 'KeyedPool' which will automatically clean up after
+-- itself when all referenced to the 'KeyedPool' are gone. It will
+-- also fork a reaper thread to regularly kill off unused resource.
+createKeyedPool
+    :: Ord key
+    => (key -> IO resource) -- ^ create a new resource
+    -> (resource -> IO ()) -- ^ destroy a resource
+    -> Int -- ^ number of resources per key to allow in the pool
+    -> Int -- ^ number of resources to allow in the pool across all keys
+    -> (SomeException -> IO ()) -- ^ what to do if the reaper throws an exception
+    -> IO (KeyedPool key resource)
+createKeyedPool create destroy maxPerKey maxTotal onReaperException = do
+    var <- newTVarIO $ PoolOpen 0 Map.empty
+
+    -- We use a different IORef for the weak ref instead of the var
+    -- above since the reaper thread will always be holding onto a
+    -- reference.
+    alive <- newIORef ()
+    mkWeakIORef alive $ destroyKeyedPool' destroy var
+
+    -- Make sure to fork _after_ we've established the mkWeakIORef. If
+    -- we did it the other way around, it would be possible for an
+    -- async exception to happen before our destroyKeyedPool' handler
+    -- was installed, and then reap would have to rely on detecting an
+    -- STM deadlock before it could ever exit. This way, the reap
+    -- function will only start running when we're guaranteed that
+    -- cleanup will be triggered.
+
+    -- Ensure that we have a normal masking state in the new thread.
+    _ <- forkIOWithUnmask $ \restore -> keepRunning $ restore $ reap destroy var
+    return KeyedPool
+        { kpCreate = create
+        , kpDestroy = destroy
+        , kpMaxPerKey = maxPerKey
+        , kpMaxTotal = maxTotal
+        , kpVar = var
+        , kpAlive = alive
+        }
+  where
+    keepRunning action =
+        loop
+      where
+        loop = action `catch` \e -> onReaperException e >> loop
+
+-- | Make a 'KeyedPool' inactive and destroy all idle resources.
+destroyKeyedPool' :: (resource -> IO ())
+                  -> TVar (PoolMap key resource)
+                  -> IO ()
+destroyKeyedPool' destroy var = do
+    m <- atomically $ swapTVar var PoolClosed
```

Comment:
I mean that the STM action may (?) block, so `atomically` could be interrupted by async exception. Then the following line will not be called at all.